File: coding-labs/low-latency-autosuggest/python/src/autosuggest_index.py
from dataclasses import dataclass
from typing import Sequence


MAX_RESULTS = 10


@dataclass(frozen=True)
class Suggestion:
    term: str
    score: int


class AutosuggestIndex:
    """Immutable top-results-per-prefix index for a bounded teaching dataset."""

    def __init__(self, suggestions: Sequence[Suggestion]):
        if suggestions is None:
            raise ValueError("suggestions are required")

        best_score_by_term: dict[str, int] = {}
        for suggestion in suggestions:
            if suggestion is None:
                raise ValueError("suggestion is required")
            if suggestion.score < 0:
                raise ValueError("score cannot be negative")
            term = self._normalize(suggestion.term)
            if term:
                best_score_by_term[term] = max(suggestion.score, best_score_by_term.get(term, -1))

        ranked = sorted(
            (Suggestion(term, score) for term, score in best_score_by_term.items()),
            key=lambda suggestion: (-suggestion.score, suggestion.term),
        )
        building: dict[str, list[Suggestion]] = {}
        for suggestion in ranked:
            self._add(building, "", suggestion)
            for end in range(1, len(suggestion.term) + 1):
                self._add(building, suggestion.term[:end], suggestion)
        self._by_prefix = {prefix: tuple(values) for prefix, values in building.items()}

    def suggest(self, prefix: str, limit: int) -> tuple[Suggestion, ...]:
        if prefix is None:
            raise ValueError("prefix is required")
        if limit < 1 or limit > MAX_RESULTS:
            raise ValueError("limit must be between 1 and 10")
        return self._by_prefix.get(self._normalize(prefix), ())[:limit]

    @staticmethod
    def _add(index: dict[str, list[Suggestion]], prefix: str, suggestion: Suggestion) -> None:
        values = index.setdefault(prefix, [])
        if len(values) < MAX_RESULTS:
            values.append(suggestion)

    @staticmethod
    def _normalize(value: str) -> str:
        if value is None:
            raise ValueError("term is required")
        return value.strip().lower()

File: coding-labs/low-latency-autosuggest/java/src/main/java/AutosuggestIndex.java
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

public final class AutosuggestIndex {
  private static final int MAX_RESULTS = 10;

  public record Suggestion(String term, long score) {}

  private final Map<String, List<Suggestion>> byPrefix;

  public AutosuggestIndex(List<Suggestion> suggestions) {
    if (suggestions == null) throw new IllegalArgumentException("suggestions are required");

    Map<String, Long> bestScoreByTerm = new HashMap<>();
    for (Suggestion suggestion : suggestions) {
      if (suggestion == null) throw new IllegalArgumentException("suggestion is required");
      if (suggestion.score() < 0) throw new IllegalArgumentException("score cannot be negative");
      String term = normalize(suggestion.term());
      if (!term.isEmpty()) bestScoreByTerm.merge(term, suggestion.score(), Math::max);
    }

    List<Suggestion> ranked = bestScoreByTerm.entrySet().stream()
        .map(entry -> new Suggestion(entry.getKey(), entry.getValue()))
        .sorted(Comparator.comparingLong(Suggestion::score).reversed().thenComparing(Suggestion::term))
        .toList();

    Map<String, List<Suggestion>> building = new HashMap<>();
    for (Suggestion suggestion : ranked) {
      add(building, "", suggestion);
      String term = suggestion.term();
      int end = 0;
      while (end < term.length()) {
        end += Character.charCount(term.codePointAt(end));
        add(building, term.substring(0, end), suggestion);
      }
    }

    Map<String, List<Suggestion>> immutable = new HashMap<>();
    building.forEach((prefix, values) -> immutable.put(prefix, List.copyOf(values)));
    byPrefix = Map.copyOf(immutable);
  }

  public List<Suggestion> suggest(String prefix, int limit) {
    if (prefix == null) throw new IllegalArgumentException("prefix is required");
    if (limit < 1 || limit > MAX_RESULTS) throw new IllegalArgumentException("limit must be between 1 and 10");
    List<Suggestion> matches = byPrefix.getOrDefault(normalize(prefix), List.of());
    return List.copyOf(matches.subList(0, Math.min(limit, matches.size())));
  }

  private static void add(Map<String, List<Suggestion>> index, String prefix, Suggestion suggestion) {
    List<Suggestion> values = index.computeIfAbsent(prefix, ignored -> new ArrayList<>());
    if (values.size() < MAX_RESULTS) values.add(suggestion);
  }

  private static String normalize(String value) {
    if (value == null) throw new IllegalArgumentException("term is required");
    return value.strip().toLowerCase(Locale.ROOT);
  }
}
